### PR TITLE
UPSTREAM: <carry>: openshift: Add unit test and only persist when needed

### PR DESCRIPTION
**Pull request by enxebre**
Add unit test and only persist when needed. We should only update the API server and increase the resource version when needed, this also prevent stale object api server updates.

### CODE REVIEW

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 	err = a.reconcilerBuilder(scope).Create(context.Background())
 	if err != nil {
 		// We still want to persist on failure to update MachineStatus
-		if err := scope.Persist(); err != nil {
+		if err := scope.PersistIfNeeded(machine.DeepCopy()); err != nil {
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator.go]: `PersistIfNeeded` doesn't mutate so the calls to `DeepCopy` are unnecessary.

File: pkg/cloud/azure/actuators/machine_scope_test.go [hunk 2]
```diff
@@ -19,14 +19,21 @@ package actuators
 import (
 	"testing"
 
+	"k8s.io/utils/pointer"
+
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine_scope_test.go]: nit: drop the blank line and accumulate with the other `k8s.io` imports.

**Comment by enxebre** [comment]
> I think we've made this worse. Why? Because scope already holds the machine to be operated on. Adding a parameter to Persist() is confusing. scope.Machine could be used or mutated. Ditto for the parameter we've just added. If scope holds the current machine then we should consistently use that.

Not sure I follow, might naming is confusing, to persist if needed you need to be aware of the non mutated version for the machine, hence the parameter to `persistIfNeeded`

**Comment by frobware** [comment]
> Not sure I follow, may be naming is confusing, to persist if needed you need to be aware of the initial non mutated version of the machine, hence the parameter to `persistIfNeeded`

And if it is updated `scope.machine` is out of date. `scope` the object is now inconsistent. Every use of scope now needs to be aware of that.

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 3]
```diff
@@ -160,26 +161,37 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
...
+	currentProviderStatus, err := v1beta1.MachineStatusFromProviderStatus(currentMachine.Status.ProviderStatus)
 	if err != nil {
-		return fmt.Errorf("[machinescope] failed to update machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+		return fmt.Errorf("failed to get machine providerStatus: %v", err)
 	}
 
-	_, err = m.storeMachineStatus(latestMachine)
-	if err != nil {
-		return fmt.Errorf("[machinescope] failed to store provider status for machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
+	modifiedMachine := s.Machine.DeepCopy()
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine_scope.go]: I think the clearest thing would be to pass `ProviderStatus` to this function (i.e., Persist). 

I find the `machine` parameter and `s.Machine` confusing. Which takes precedence? Should the machine object we're trying to modify be `machine` or `s.Machine`? It seems that we're only going to do some work if `ProviderStatus` is different - so we could narrow down to just that. 

Thoughts?

Comment (enxebre) [reply pkg/cloud/azure/actuators/machine_scope.go]: currentMachine parameter is the original non mutated object so you can check against it to decided if you want to update and hit the API server (may be naming is not clear)
> It seems that we're only going to do some work if ProviderStatus

This needs to check the machine.Spec, machine.Spec.ProviderSpec, machine.Status and machine.Status.ProviderStatus

Comment (frobware) [reply pkg/cloud/azure/actuators/machine_scope.go]: > This needs to check the machine.Spec, machine.Spec.ProviderSpec, machine.Status and machine.Status.ProviderStatus

OK, so pass those things.

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 4]
```diff
@@ -160,26 +161,37 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
...
+	if !equality.Semantic.DeepEqual(currentMachine, modifiedMachine) ||
+		!equality.Semantic.DeepEqual(currentConfig, s.MachineConfig) {
+		modifiedMachine, err = s.storeMachineSpec(modifiedMachine)
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine_scope.go]: IIUC, the machine object is updated to reflect the new machine spec (assuming `storeMachineSpec` is successful) but we don't refelct that in `s.Machine`. If you now go on to use `s.Machine` (here, or from the callers perspective) then this is now out of date.

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 5]
```diff
@@ -114,7 +114,7 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 		}
 	}
 
-	if err := scope.Persist(); err != nil {
+	if err := scope.PersistIfNeeded(machine.DeepCopy()); err != nil {
```
Comment (ingvagabund) [inline pkg/cloud/azure/actuators/machine/actuator.go]: Why is `machine.DeepCopy` here and `machine` everywhere else?

**Comment by ingvagabund** [comment]
I don't see a point of fixing scope at all. Too many open wounds. I would rather drop the machine scope (or any scope) completely and do what we do in our aws actuator.

Feel free to unhold the PR though it's no-go for me.

/hold

File: pkg/cloud/azure/actuators/machine_scope.go (hunk 4 above)
Comment (frobware) [reply pkg/cloud/azure/actuators/machine_scope.go]: We should reflect `modifiedMachine` into `s.Machine` (using `DeepCopyInto`). If we do that I think I can live with the rest.

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 6]
```diff
@@ -160,26 +161,64 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
...
+	}
+	currentProviderStatus, err := v1beta1.MachineStatusFromProviderStatus(currentMachine.Status.ProviderStatus)
+	if err != nil {
+		return fmt.Errorf("failed to get machine providerStatus: %v", err)
+	}
+
+	// update machine if needed
+	if !equality.Semantic.DeepEqual(currentMachine, s.Machine) ||
+		!equality.Semantic.DeepEqual(currentConfig, s.MachineConfig) {
+		latestMachine, err := s.storeMachineSpec(s.Machine.DeepCopy())
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine_scope.go]: shouldn't this pass `currentMachine` to `storeMachineSpec`?

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 7]
```diff
@@ -160,26 +161,64 @@ func (m *MachineScope) storeMachineStatus(machine *machinev1.Machine) (*machinev
...
+		if err := s.setScopeValuesFromMachine(latestMachine); err != nil {
+			return fmt.Errorf("failed to set new values from machine: %v", err)
+		}
 	}
 
-	latestMachine, err := m.storeMachineSpec(m.Machine)
+	// update status if needed
+	if !equality.Semantic.DeepEqual(currentMachine.Status, s.Machine.Status) ||
+		!equality.Semantic.DeepEqual(currentProviderStatus, s.MachineStatus) {
+		latestMachine, err := s.storeMachineStatus(s.Machine.DeepCopy())
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine_scope.go]: shouldn't this pass `currentMachine` to `storeMachineStatus`?

**Pull request closed without merging**